Add Micronaut discovery core optional dependency to discovery features

wetted opened the pull request:
This depends on https://github.com/micronaut-projects/micronaut-starter/pull/1515 before it will build successfully, but it cannot be a pending feature for 3.x because it has breaking changes for an existing feature.

closes #1616

wetted committed:
Micronaut discovery core is now an optional dependency and must be added for Starter discovery features.

closes #1616

wetted committed:
Micronaut discovery core is now an optional dependency and must be added for Starter discovery features.

closes #1616

sdelamo commented on starter-core/src/test/groovy/io/micronaut/starter/feature/discovery/DiscoveryConsulSpec.groovy:
```diff
@@ -31,6 +31,7 @@ class DiscoveryConsulSpec extends ApplicationContextSpec  implements CommandOutp
                 .render()
 
         then:
+        template.contains('implementation("io.micronaut:micronaut-discovery-core")')
```
Use BuildVerifier API to test dependencies.

wetted replied:
@sdelamo The BuildTestVerifier API is in 3.9.x but not 4.0.x, which this PR has to target. Do you want to merge it up?

sdelamo commented on starter-core/src/test/groovy/io/micronaut/starter/feature/discovery/DiscoveryConsulSpec.groovy:
```diff
@@ -53,6 +54,13 @@ class DiscoveryConsulSpec extends ApplicationContextSpec  implements CommandOutp
       <scope>compile</scope>
     </dependency>
 """)
+        template.contains("""
```
use BuildVerifierApi to test dependencies

sdelamo commented on starter-core/src/test/groovy/io/micronaut/starter/feature/discovery/DiscoveryKubernetesSpec.groovy:
```diff
@@ -29,6 +29,7 @@ class DiscoveryKubernetesSpec extends ApplicationContextSpec implements CommandO
                 .render()
 
         then:
+        template.contains('implementation("io.micronaut:micronaut-discovery-core")')
```
Use BuildVerifier API to tests dependencies

sdelamo commented on starter-core/src/test/groovy/io/micronaut/starter/feature/discovery/DiscoveryKubernetesSpec.groovy:
```diff
@@ -51,6 +52,13 @@ class DiscoveryKubernetesSpec extends ApplicationContextSpec implements CommandO
       <scope>compile</scope>
     </dependency>
 """)
+        template.contains("""
```
Use BuildVerifier API to tests dependencies

sdelamo commented on starter-core/src/test/groovy/io/micronaut/starter/feature/discovery/EurekaSpec.groovy:
```diff
@@ -29,6 +29,7 @@ class EurekaSpec extends ApplicationContextSpec implements CommandOutputFixture
                 .render()
 
         then:
+        template.contains('implementation("io.micronaut:micronaut-discovery-core")')
```
Use BuildVerifier API to tests dependencies

sdelamo commented on starter-core/src/test/groovy/io/micronaut/starter/feature/discovery/EurekaSpec.groovy:
```diff
@@ -51,6 +52,13 @@ class EurekaSpec extends ApplicationContextSpec implements CommandOutputFixture
       <scope>compile</scope>
     </dependency>
 """)
+        template.contains("""
```
Use BuildVerifier API to tests dependencies

sdelamo commented on starter-core/src/main/java/io/micronaut/starter/feature/discovery/Eureka.java:
```diff
@@ -40,6 +39,7 @@ public String getDescription() {
 
     @Override
     public void apply(GeneratorContext generatorContext) {
+        applyCore(generatorContext);
```
```suggestion
        super.apply(generatorContext);
```

sdelamo commented on starter-core/src/main/java/io/micronaut/starter/feature/discovery/DiscoveryKubernetes.java:
```diff
@@ -44,6 +44,7 @@ public String getDescription() {
     public void apply(GeneratorContext generatorContext) {
         generatorContext.getBootstrapConfiguration().put("kubernetes.client.discovery.mode", "endpoint");
         generatorContext.getBootstrapConfiguration().put("kubernetes.client.discovery.mode-configuration.endpoint.watch.enabled", true);
+        applyCore(generatorContext);
```
```suggestion
        super.apply(generatorContext);
```

sdelamo commented on starter-core/src/main/java/io/micronaut/starter/feature/discovery/DiscoveryCore.java:
```diff
@@ -0,0 +1,27 @@
+import io.micronaut.starter.build.dependencies.MicronautDependencyUtils;
+
+public abstract class DiscoveryCore implements DiscoveryFeature {
+    protected void applyCore(GeneratorContext generatorContext) {
```
```suggestion
    public void apply(GeneratorContext generatorContext) {
```

sdelamo commented on starter-core/src/main/java/io/micronaut/starter/feature/discovery/DiscoveryCore.java:
```diff
@@ -0,0 +1,27 @@
+
+public abstract class DiscoveryCore implements DiscoveryFeature {
+    protected void applyCore(GeneratorContext generatorContext) {
+        generatorContext.addDependency(MicronautDependencyUtils.coreDependency()
```
extract constant: 

```java
private final static Dependency DEPENDENCY_MICRONAUT_DISCOVERY_CORE = MicronautDependencyUtils.coreDependency()
                 .artifactId("micronaut-discovery-core")
                 .compile()
                 .build();
```

sdelamo commented on starter-core/src/main/java/io/micronaut/starter/feature/discovery/DiscoveryCore.java:
```diff
@@ -0,0 +1,27 @@
+import io.micronaut.starter.application.generator.GeneratorContext;
+import io.micronaut.starter.build.dependencies.MicronautDependencyUtils;
+
+public abstract class DiscoveryCore implements DiscoveryFeature {
```
I think we will need a non abstract class so that users are able to select a discovery-core even if they don't use consul, k8s or eureka.

wetted committed:
Bump views-jte gradle and maven plugins to 2.2.6 (#1646)

* Bump views-jte gradle and maven plugins to 2.3.0

fixes #1645

* Downgrade to jte plugin version 2.2.6. It's the latest we can use for Micronaut 3.x since jte plugin 2.3.0 uses Kotlin 1.8.x which causes build model failures.

sdelamo committed:
Merge branch '3.8.x' into 3.9.x

sdelamo committed:
flyway delete datasource.default.schema-generate

wetted committed:
Upgrading to Maven 3.9.0 for maven applications generated with Micronaut CLI/Launch (#1617)

* Upgrading to Maven 3.9.0 for maven applications generated with Micronaut CLI/Launch

closes #1606

* remove MavenWrapperDownloader

sdelamo committed:
build: Micronaut Framework 3.8.7

micronaut-build committed:
[skip ci] Release v3.8.7

micronaut-build committed:
Back to 3.8.8-SNAPSHOT

sdelamo committed:
build: Micronaut Framework 3.8.8-SNAPSHOT

sdelamo committed:
Merge branch '3.8.x' into 3.9.x

wetted committed:
Bump kotlin from 1.6.10 to 1.6.21. (#1648)

* Bump kotlin from 1.6.10 to 1.6.21. This is what the 3.8.x branch has, and is required for `views-jte` version of the jte gradle plugin it uses.

* extract java home variable [ci skip]

* Fix version in TOML [ci skip]

---------

Co-authored-by: sdelamo <sergio.delamo@softamo.com>
Co-authored-by: Tim Yates <tim.yates@gmail.com>

wetted committed:
AWS Lambda using Java runtime enables SnapStart by default (#1642)

alvarosanchez committed:
Fix AOT Maven feature (#1649)

guillermocalvo committed:
Add feature for google cloudevents function (#1550)

* Add new feature for GCP CloudEvents functions

* Add tests

* Add one more unit test for the feature

* Use `MicronautDependencyUtils` instead of `Dependency.builder()`

* Use `BuildTestVerifier` to test dependencies in Gradle/Maven

* Update test-cloud/src/test/groovy/io/micronaut/starter/core/test/cloud/gcp/CreateGoogleCloudFunctionCloudEventsSpec.groovy

* extract dependency constants

Test feature supports

---------

Co-authored-by: Sergio del Amo <sergio.delamo@softamo.com>

musketyr committed:
feat Micronaut Slack (#1458)

melix committed:
Upgrade Gradle plugin version (#1653)

sdelamo committed:
feat: support for Gradle Enterprise for Maven builds (#1595)

radovanradic committed:
Feature for data-azure-cosmos (#1584)

wetted committed:
Fix log4j2 feature generating incorrect log4j2 dependencies (#1583)

sdelamo committed:
Merge branch '3.8.x' into 3.9.x

wetted committed:
Add `micronaut-aop` feature. (#1652)

sdelamo committed:
add utility method valuesGradle() (#1657)

sdelamo committed:
allow features to contribute repositories (#1658)

timyates committed:
Update Spock to the latest (#1665)

* Update Spock to the latest

This allows us to use new features such as lenient comparison instead of comparing sets

https://spockframework.org/spock/docs/2.3/utilities.html\#collection-conditions

* Use set comparison in BuildToolSpec

sdelamo committed:
support api gradle configuration (#1659)

sdelamo committed:
GraalVMFeatureValidator::supports) (#1660)

sdelamo committed:
MicronautDependencyUtils gcp (#1661)

sdelamo committed:
support azurefunctions with Gradle DSL (#1663)

sdelamo committed:
add default interfaces for clouds (#1664)

sdelamo committed:
Add method hasBuildDependency to BuildVerifier (#1662)

* hasBuildDependency

* rename to hasBuildPlugin

* use BuildTestVerifier::hasBuildPlugin

sdelamo committed:
AWS Lambda Dependencies (#1669)

sdelamo committed:
MicronautRuntimeFeature improvements [ci skip] (#1667)

timyates committed:
Gradle repository mavenContent filters (#1670)

We had these for Maven, but not Gradle.  This brings the two into line

sdelamo committed:
azure Gradle plugin and host.json local.setting.json to match values in sample Azure Functions application (#1666)

sdelamo committed:
extract HandlerProvider (#1668)

timyates committed:
Switch default JDK in Micronaut Launch to 17 (#1672)

timyates committed:
Show the available JDK options in the CLI --help (#1673)

* Show the available JDK options in the CLI --help

I was going to go further, and show the default, but the default for AWS is 11 (depending on service).
Also, the default language changes the default gradle dsl and the default test framework.
This means I don't believe there's much more we can do of value.

* Add typehint for native compilation

sdelamo committed:
add crac dependency for lambda and snapstart (#1676)

graemerocher committed:
Build a fully static Linux Image (#1685)

wetted committed:
Adding README.md links to docs for build plugins (gradle, graalvm, maven) (#1684)

* Adding README.md links to docs for build plugins (gradle, graalvm, maven).

closes #1682

* Use `GraalVMFeatureValidator` to test for including graalvm plugin docs.

sdelamo committed:
build: bump Micronaut gradle plugin to 3.7.7 (#1686)

https://github.com/micronaut-projects/micronaut-gradle-plugin/releases/tag/v3.7.7

https://github.com/micronaut-projects/micronaut-gradle-plugin/releases/tag/v3.7.6

https://github.com/micronaut-projects/micronaut-gradle-plugin/releases/tag/v3.7.5

grimmjo committed:
fixes default nats-configuration and adds link to documentation (#1607)

sdelamo committed:
build: Micronaut Framework 3.8.8

micronaut-build committed:
[skip ci] Release v3.8.8

micronaut-build committed:
Back to 3.8.9-SNAPSHOT

sdelamo committed:
Micronaut Framework 3.8.9-SNAPSHOT

sdelamo committed:
Merge branch '3.8.x' into 3.9.x

# Conflicts:
#	gradle.properties
#	settings.gradle
#	starter-core/src/main/resources/pom.xml
#	starter-core/src/test/groovy/io/micronaut/starter/feature/build/MicronautAotSpec.groovy

wetted committed:
The new syntax for setting output values requires changing `$GITHUB_OUTPUT` to `$env:GITHUB_OUTPUT` for steps that run in Windows Powershell. See https://stackoverflow.com/a/74444094/6509 (#1688)

fixes #1687

wetted committed:
Merge remote-tracking branch 'origin/4.0.x' into add_discovery_core_artifact

wetted committed:
[skip ci]
- use BuildVerifier for tests
- change DiscoveryCore feature to non-abstract
- other changes to address review feedback

wetted committed:
[skip ci]
- use BuildVerifier for tests
- change DiscoveryCore feature to non-abstract
- other changes to address review feedback

wetted committed:
Maven wrapper to Maven 3.9.1 (#1694)

closes #1693

wetted committed:
Micronaut Session has different coordinates in Micronaut Framework 4.0.x (#1624)

* Micronaut Session has different coordinates in Micronaut Framework 4.0. This fixes the HttpSession feature for that.

closes #1615

* [skip ci] use BuildVerifier for tests

sdelamo committed:
add ability to contribute Gradle repositories to settings (#1675)

tobiasschaefer committed:
Add configuration to camunda-platform7 and update metadata of Camunda related projects (#1689)

timyates committed:
Gradle 7.6.1 (#1552)

jdubois committed:
Upgrade to PostgreSQL 14 for TestContainers (#1695)

sdelamo committed:
Merge branch '3.8.x' into 3.9.x

sdelamo committed:
pin baseImage to amazonlinux:2 (#1701)

sdelamo committed:
pin baseImage to amazonlinux:2 (#1701)

sdelamo committed:
don't offer function-url when selecting controllers style (#1703)

sdelamo committed:
don't offer function-url when selecting controllers style (#1703)

wetted committed:
AWS-lambda features without GraalVM feature raises validation error when Java 17 is selected (#1700)

* AWS-lambda features without Graalvm should throw an error if Java 17 is selected.

closes #1699

* Disable tests for JDK 17 that use AWS Lambda features without Graalvm feature, a combination no longer allowed with AwsLambdaFeatureValidator.

Smell: some of these tests are not directly related to testing `AwsLambda` features, but merely test something else for all application types, where `ApplicationType.FUNCTION` by default selects the "aws-lambda" feature, that now results in raising an  exception for JDK 17. With Micronaut 3.x these tests will run when CI is Jdk 8 and Jdk 11. However, with Micronaut 4 they won't run at all. Maybe an alternative is to rewrite those tests to either include the "graalvm" feature or a Jdk 17 FUNCTION supported feature )OCI or Azure).

* remove @IgnoreIf

* set jdk11 as max version

---------

Co-authored-by: Sergio del Amo <sergio.delamo@softamo.com>

wetted committed:
Projects with Maven/Koltin/GraalVM features had `micronaut-graal` added to kapt plugin annotationProcessors three times.

wetted committed:
fix checkstyle errors

wetted committed:
remove `org.graalvm.nativeimage` from rocker template (it's in feature class now)

sdelamo committed:
build: Micronaut Gradle Plugin 3.7.8 (#1705)

sdelamo committed:
build: Micronaut Gradle Plugin 3.7.8 (#1705)

sdelamo committed:
build: Micronaut Framework 3.8.9

sdelamo committed:
build: Micronaut Framework 3.8.9

sdelamo committed:
ci: move shell pwsh to choco step [ci skip]

this was incorrect https://github.com/micronaut-projects/micronaut-starter/pull/1688

sdelamo committed:
ci: move shell pwsh to choco step [ci skip]

this was incorrect https://github.com/micronaut-projects/micronaut-starter/pull/1688

micronaut-build committed:
[skip ci] Release v3.8.9

micronaut-build committed:
[skip ci] Release v3.8.9

micronaut-build committed:
Back to 3.8.10-SNAPSHOT

micronaut-build committed:
Back to 3.8.10-SNAPSHOT

sdelamo committed:
back to snapshot

sdelamo committed:
back to snapshot

micronaut-build committed:
Bump Micronaut Chocolatey package to v3.8.9

wetted committed:
Merge pull request #1706 from micronaut-projects/cleanup_kotlin_graal_maven_build

Cleanup generation of maven-kotlin-graalvm projects

sdelamo committed:
remove empty lines

sdelamo committed:
Merge branch '3.8.x' into 3.9.x

wetted committed:
Fix missing dependencies for Gradle projects with "oracle-function-http" feature, (#1709)

* Fixing missing dependencies for Gradle projects with "oracle-function-http" feature, to make consistent with Maven projects. Refactor dependency logic from Rocker files to Feature classes.

* make regex group non-capturing as it was originally

* add test GradleBuildTestVerifier

* don't add micronaut-oraclecloud-function-http-test micronaut-oraclecloud-function-http for gradle

---------

Co-authored-by: Sergio del Amo <sergio.delamo@softamo.com>

wetted committed:
Fixing missing dependencies for Gradle projects with "azure-function-http" feature (#1708)

* Fixing missing dependencies for Gradle projects with "azure-function-http" feature, to make constant with Maven projects. Refactor dependency logic from Rocker files to Feature classes.

* fic checkstyle error

* Add `micronaut-azure-function` only for FUNCTION app type.

* add dependencies only for maven

Gralde adds those dependencies automatically https://github.com/micronaut-projects/micronaut-gradle-plugin/blob/3.7.x/minimal-plugin/src/main/java/io/micronaut/gradle/MicronautRuntimeDependencies.java

* set previous behaviour

---------

Co-authored-by: Sergio del Amo <sergio.delamo@softamo.com>

sdelamo committed:
build: Micronaut Gradle Plugin 3.7.8

sdelamo committed:
Micronaut Build Plugins 5.4.8

sdelamo committed:
Micronaut Framework 3.9.0

micronaut-build committed:
[skip ci] Release v3.9.0

micronaut-build committed:
Back to 3.9.1-SNAPSHOT

sdelamo committed:
ci: projectVersion=3.10.0-SNAPSHOT

sdelamo committed:
ci: micronaut=3.10.0-SNAPSHOT

sdelamo committed:
pulbish snapshots for 3.10.x

sdelamo committed:
build: Micronaut Framework 3.9.1-SNAPSHOT

sdelamo committed:
fix: use allowed-origins-regex configuration

sdelamo committed:
build: groovy version 4.0.11

sdelamo committed:
Merge branch '3.10.x' into 4.0.x

tobiasschaefer committed:
Update Camunda projects to newest versions (#1720)

wetted committed:
Refactor rocker dependencies into java Feature API (partial) (#1707)

* Refactor `jackson-module-kotlin` dependency from rocker templates to feature class.

* Refactor `neo4j-harness` dependency from rocker templates to Neo4JBolt class.

* Refactor `micronaut-servlet-core` dependency for "google-cloud-function-*" from rocker templates to feature classes.

* Refactor maven + groovy lang dependencies from pom rocker template to `Groovy` feature class.

* Refactor `micronaut-function` dependency from gradle rocker file for aws-lambda, aws-lambda-custom-runtime to feature classes. Maven build includes it in parent pom.

* Refactor maven + groovy `micronaut-openapi` dependency from pom rocker template to `Openapi` feature class.

* fix dependency notation error for jackson-module-kotlin

* address review feedback

* [skip CI] Refactor GCP dependencies from Gradle/Maven rocket files into GCP feature classes and add tests

* Refactor `micronaut-inject` dependency from Maven pom rocker file to build feature class. [skip ci]

* fix test dependency for gradle builds only

* use BuildTestVerifier

* add test for jackson-module-kotlin

* add comment

* test gcp-function & spock needs mn-servlet-core

* no need for unecessary if

* remove unnecessary asserts

* open api test

* add functions-framework-api as provided to maven

---------

Co-authored-by: Sergio del Amo <sergio.delamo@softamo.com>

wetted committed:
Add feature class for Micronaut Validation (#1717)

* Add feature class for Micronaut Validation and move dependency from rockers files to Feature class. [skip ci]

* Make Micronaut Validation a one-of feature with Hibernate validator and update tests. Refactor validation features into new package.

* Changes to address review feedback.

* fix checkstyle failure

* move back to other package

---------

Co-authored-by: Sergio del Amo <sergio.delamo@softamo.com>

sdelamo committed:
remove http-client from rocker (#1725)

Close #1719

sdelamo committed:
fix rocker template

sdelamo committed:
Merge branch '3.9.x' into 3.10.x

sdelamo committed:
Merge branch '3.10.x' into 4.0.x

sdelamo committed:
Micronaut Framework 4.0.x  (#1723)

graemerocher committed:
Add KSP feature (#1704)

sdelamo committed:
Run Maven tests (#1727)

* run maven tests

* define micronaut.core.version for micronaut-graal

sdelamo committed:
disable Jansi (#1729)

wetted committed:
Merge remote-tracking branch 'origin/add_discovery_core_artifact' into add_discovery_core_artifact

wetted commented:
I screwed this branch up. I'll resubmit on another PR. 🥴
